### ISSUE TITLE
Inequality constraints for Unknowns

### ISSUE THREAD

**Issue by iraikov** [tshort/Sims.jl#32]
The IDA solver provides the option of setting inequality constraints on the computed solution (i.e. y[1] >= 0). Is it possible to extend the definition of Unknown to allow such constraints? E.g. Unknown(1.0,"x",constraint=Positive)
or Unknown(1.0,"x",constraint=Negative) or similar.

**Comment by tshort** [comment]
+1. 

DASSL also supports this.

**Comment by iraikov** [comment]
Do you have any preference for the syntax? Perhaps the simplest is to use symbols for the different flags, eg. :Positive, :ZeroPositive, :Negative, :ZeroNegative  but if there is a more idiomatic way I am open to suggestions.

**Comment by tshort** [comment]
Symbols are good. I'd stick to lower case.

**Comment by tshort** [comment]
We could also consider putting the constraint flag in as a type parameter for Unknowns. Then, modelers could dispatch on that Unknown type and set up aliases. It might look like this:
```julia
typealias Voltage Unknown{UVoltage, Normal}       # no constraint
typealias Length Unknown{ULength, NonNegative}    # zero or positive
```

**Comment by iraikov** [comment]
I really like the idea of putting the constraint in the type. Would the following code work?

```julia
abstract UnknownConstraint
type Negative <: UnknownConstraint
end
type NonNegative <: UnknownConstraint
end
type Positive <: UnknownConstraint
end
type NonPositive <: UnknownConstraint
end

type Unknown{T<:UnknownCategory,C<:UnknownConstraint} <: UnknownVariable
...

```

**Comment by tshort** [comment]
That should work, @iraikov!

**Comment by iraikov** [comment]
I created a variant of type Unknown as above, but I am having a difficult time created a constructor that accepts UnknownConstraint type as an argument. I have the following code currently:
```julia

type Unknown{T<:UnknownCategory,C<:UnknownConstraint} <: UnknownVariable
...

Unknown(value = 0.0, label::String = "", fixed::Bool = false, save_history::Bool = true, constraint::UnknownConstraint = Normal) =
    Unknown{DefaultUnknown,constraint}(value, label, fixed, save_history)
```

However, it seems that I cannot pass a subtype of UnknownConstraint to the constructor. What is the proper way to do this?

**Comment by tshort** [comment]
Do you have the code in a branch somewhere?

**Comment by iraikov** [comment]
Hi Tom,

The code is here, thanks for looking at it:

https://github.com/iraikov/Sims.jl/tree/constraints

**Comment by tshort** [comment]
Looks like your code works to me. You might have problems with other arguments. Here's a log:

```julia
julia> Unknown{DefaultUnknown, Positive}()
<<`8173`,0.0>>

julia> Unknown{DefaultUnknown, Positive}(1.2)
<<`8174`,1.2>>

julia> Unknown(1.2)
<<`8175`,1.2>>

julia> dump(Unknown(1.2))
Unknown{DefaultUnknown,Normal}
  sym: Symbol ##8176
  value: Float64 1.2
  label: ASCIIString ""
  fixed: Bool false
  save_history: Bool true

julia> t(constraint) = Unknown{DefaultUnknown,constraint}()
t (generic function with 1 method)

julia> t(Normal)
<<`8177`,0.0>>

julia> t(Positive)
<<`8178`,0.0>>

julia> t(NonPositive)
<<`8179`,0.0>>

julia> NonPositive
NonPositive (constructor with 1 method)

julia> t(;constraint = Negative) = Unknown{DefaultUnknown,constraint}()
t (generic function with 2 methods)

julia> t()
<<`8180`,0.0>>

julia> t(constraint = Positive)
<<`8181`,0.0>>
```

**Comment by tshort** [comment]
I think I see it. You probably need to export UnknownConstraint.

**Comment by tshort** [comment]
@iraikov, in your branch, you've defined several new Unknown method definitions, like NonPositiveUnknown as well as an option to specify the constraint with a keyword argument. I think that's too many ways to do the same thing. I'd prefer just to stick to Unknown{type, constraint}(). I think the constraint will mostly be used with type aliases.

**Comment by iraikov** [comment]
Hi Tom,

You are right, I realized I had 'fixed' the issue I had by creating one
constructor for each constraint. Yesterday, I pressed forward with the
interface to the solvers to see how things will go, but I will remove all
the extra constructor and try again Unknown{type,constraint} and
typealiases, as you suggest. Unfortunately, so far it seems that dassl
cannot handle some systems with constraints which it can solve without the
constraints. Would you mind having a look at my modifications to dassl on
that branch?

**Comment by tshort** [comment]
I took a quick look at the dassl code. It looks good to me. Maybe sundials will work better.

**Comment by tshort** [comment]
Closed by PR #48.
